File: rust/src/p0020.rs
/*
Project Euler Problem 20

Problem:

n! means n × (n − 1) × ... × 3 × 2 × 1

For example, 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800,
and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27.

Find the sum of the digits in the number 100!
*/

pub fn p0020() -> i128 {
    let mut numbers: Vec<u128> = vec![0; 5];
    let ten36: u128 = 1000000000000000000000000000000000000;
    numbers[0] = 1;
    for i in 2..=100 {
        for number in numbers {
            number *= i;
        }
        for j in 0..4 {
            if numbers[j] > ten36 {
                numbers[j + 1] += numbers[j] / ten36;
                numbers[j] %= ten36;
            }
        }
    }
    let mut answer: i128 = 0;
    let mut power: u128 = 1;
    for _ in 0..36 {
        for number in numbers {
            answer += ((number / power) % 10) as i128;
        }
        power *= 10;
    }
    return answer;
}

File: javascript/src/p0020.js
/**
 * Project Euler Problem 20
 *
 * Problem:
 * @return {number}
 */
exports.p0020 = function() {
    const ten13 = 10000000000000;
    const numbers = [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
    for (let i = 2; i <= 100; i++) {
        for (let j = 0; j < numbers.length; j++) {
            numbers[j] *= i;
        }
        for (let j = 0; j < numbers.length - 1; j++) {
            if (numbers[j] > ten13) {
                numbers[j + 1] += 0 | (numbers[j] / ten13);
                numbers[j] %= ten13;
            }
        }
    }
    let answer = 0;
    let power = 1;
    while (power < ten13) {
        for (let j = 0; j < numbers.length; j++) {
            answer += 0 | ((numbers[j] / power) % 10);
        }
        power *= 10;
    }
    return answer;
};
